src/v/cloud_storage/remote_partition.cc

```diff
@@ -0,0 +1,652 @@
+        co_await _reader->stop();
+        _it = _partition->_segments.end();
+        _reader = {};
+    }
+
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+
+    ss::lw_shared_ptr<remote_partition> _partition;
+    /// Currently accessed segment
+    remote_partition::segment_map_t::iterator _it;
+    /// Reader state that was borrowed from the materialized_segment_state
+    std::unique_ptr<remote_segment_batch_reader> _reader;
+    /// Cancelation subscription
+    ss::abort_source::subscription _as_sub;
+};
+
+remote_partition::remote_partition(
+  const manifest& m, remote& api, cache& c, s3::bucket_name bucket)
+  : _rtc()
+  , _ctxlog(cst_log, _rtc, m.get_ntp().path())
+  , _api(api)
+  , _cache(c)
+  , _manifest(m)
+  , _bucket(std::move(bucket))
+  , _stm_jitter(stm_jitter_duration) {}
+
+ss::future<> remote_partition::start() {
+    update_segments_incrementally();
+    (void)run_eviction_loop();
+
+    _stm_timer.set_callback([this] {
+        gc_stale_materialized_segments();
+        if (!_as.abort_requested()) {
+            _stm_timer.rearm(_stm_jitter());
+        }
+    });
+    _stm_timer.rearm(_stm_jitter());
+    co_return;
+}
+
+ss::future<> remote_partition::run_eviction_loop() {
+    // Evict readers asynchronously
+    gate_guard g(_gate);
+    try {
+        while (!_as.abort_requested()) {
+            co_await _cvar.wait([this] {
+                return _as.abort_requested() || !_eviction_list.empty();
+            });
+            auto tmp_list = std::exchange(_eviction_list, {});
+            for (auto& rs : tmp_list) {
+                co_await std::visit([](auto&& rs) { return rs->stop(); }, rs);
+            }
+        }
+    } catch (const ss::broken_condition_variable&) {
+    }
+    vlog(_ctxlog.debug, "remote partition eviction loop stopped");
+}
+
+void remote_partition::start_readahead(
+  remote_partition::segment_map_t::iterator current) {
+    struct visit_hydrate {
+        segment_state& state;
+        remote_partition* part;
+        const model::offset offset_key;
+
+        void operator()(offloaded_segment_ptr& st) {
+            vlog(
+              part->_ctxlog.debug,
+              "remote partition readahead, hydrating {}",
+              st->manifest_key);
+            auto tmp = st->materialize(*part, offset_key);
+            (void)tmp->segment->hydrate().discard_result().handle_exception(
+              [this](const std::exception_ptr& e) {
+                  vlog(
+                    part->_ctxlog.error,
+                    "Error {} while prefetching segment",
+                    e);
+              });
+            state = std::move(tmp);
+        }
+        void operator()(materialized_segment_ptr&) {
+            // The segment was accessed recently so it should be already
+            // hydrated
+        }
+    };
+    // prefetch if on the segment's boundary
+    auto pi = current;
+    auto readahead = stm_readahead;
+    while (readahead--) {
```

Comment:
Looks like readahead is a no-op right now. Maybe we could delete the code or leave a TODO comment on what needs to be done to enable it again?